test/libsolidity/SyntaxTest.cpp

```diff
@@ -80,34 +81,28 @@ void SyntaxTest::setupCompiler(CompilerStack& _compiler)
 
 void SyntaxTest::parseAndAnalyze()
 {
-	try
-	{
-		runFramework(withPreamble(m_sources.sources), m_stopAfter);
-		if (!pipelineSuccessful() && stageSuccessful(PipelineStage::Analysis) && !compiler().isExperimentalAnalysis())
-		{
-			ErrorList const& errors = compiler().errors();
-			auto codeGeneretionErrorCount = count_if(errors.cbegin(), errors.cend(), [](auto const& error) {
-				return error->type() == Error::Type::CodeGenerationError;
-			});
-			auto errorCount = count_if(errors.cbegin(), errors.cend(), [](auto const& error) {
-				return Error::isError(error->type());
-			});
-			// failing compilation after successful analysis is a rare case,
-			// it assumes that errors contain exactly one error, and the error is of type Error::Type::CodeGenerationError
-			if (codeGeneretionErrorCount != 1 || errorCount != 1)
-				BOOST_THROW_EXCEPTION(std::runtime_error("Compilation failed even though analysis was successful."));
-		}
-	}
-	catch (UnimplementedFeatureError const& _e)
+
+	runFramework(withPreamble(m_sources.sources), m_stopAfter);
+	if (!pipelineSuccessful() && stageSuccessful(PipelineStage::Analysis) && !compiler().isExperimentalAnalysis())
 	{
-		m_errorList.emplace_back(SyntaxTestError{
-			Error::Type::UnimplementedFeatureError,
-			std::nullopt,
-			errorMessage(_e),
-			"",
-			-1,
-			-1
-		});
+		ErrorList const& errors = compiler().errors();
+		static auto allowedErrors = [](std::shared_ptr<Error const> const& _error) {
```

Comment:
Actually, the initial suggestion of `isICE` (or `isInternalError`) was better here:

```suggestion
		static auto isInternalError = [](std::shared_ptr<Error const> const& _error) {
```

I'd also adjust the comment below to say:
```c++
// Most errors are detected during analysis, and should not happen during code generation.
// There are some exceptions, e.g. unimplemented features or stack too deep, but anything else at this stage
// is an internal error that signals a bug in the compiler (rather than in user's code).
```

Reply:
Done.